benchmark/benchmain/main.go

```diff
@@ -106,6 +106,10 @@ var (
 	useBufconn          = flag.Bool("bufconn", false, "Use in-memory connection instead of system network I/O")
 	enableKeepalive     = flag.Bool("enable_keepalive", false, "Enable client keepalive. \n"+
 		"Keepalive.Time is set to 10s, Keepalive.Timeout is set to 1s, Keepalive.PermitWithoutStream is set to true.")
+	clientReadBufferSize  = flags.IntSlice("clientReadBufferSize", []int{-1}, "Configures the client read buffer size in bytes. If negative, use the default - may be a a comma-separated list")
```

Comment:
For these flags to work when specified as a comma separated list, don't we have to add corresponding fields to `featureOpts` struct, and generate appropriate `stats.Feature` in the `generateFeatures` function?

I agree that this is not an easy to read code. It pre-dates most of us on the team, and at some point I tried to clean this benchmark code a little bit. Didn't succeed a whole lot I guess.

We have two options:
- Make these flags to be of type `int`, and therefore accept only one value (this means we would have to run the benchmark multiple times for different buffer sizes, which might not be all that bad)
- Or keep them as `IntSlice` and ensure that they are plumbed all the way through like other `IntSlice` flags.

On another note, I just realized that we are not sanitizing the buffer sizes received through the dial options. Passing negative values would cause a panic I think. So, I would advise against using a negative value for the flag to indicate the use of the default value.

Reply:
> For these flags to work when specified as a comma separated list, don't we have to add corresponding fields to featureOpts struct, and generate appropriate stats.Feature in the generateFeatures function?
> I agree that this is not an easy to read code. It pre-dates most of us on the team, and at some point I tried to clean this benchmark code a little bit. Didn't succeed a whole lot I guess.

Of course you're right :). The code might be a bit heavy for what it's doing, but I certainly should have tested that my change works properly -- apologize for that. I fixed that in a new commit.

> We have two options:

> Make these flags to be of type int, and therefore accept only one value (this means we would have to run the benchmark multiple times for different buffer sizes, which might not be all that bad)
> Or keep them as IntSlice and ensure that they are plumbed all the way through like other IntSlice flags.

I chose the second option of plumbing int slices properly. Let me know if you prefer the other option. Here is the result:

```
$ go run benchmark/benchmain/main.go -reqSizeBytes 4096 -respSizeBytes 4096 -respSizeBytes 0 -kbps 0 -workloads unary -latency 0 -maxConcurrentCalls 10 -serverReadBufferSize -1 -serverWriteBufferSize 0,65536 -clientReadBufferSize -1 -clientWriteBufferSize -1
go1.19.1/grpc1.51.0-dev
unary-networkMode_none-bufConn_false-keepalive_false-benchTime_1s-trace_false-latency_0s-kbps_0-MTU_0-maxConcurrentCalls_10-reqSize_4096B-respSize_0B-compressor_off-channelz_false-preloader_false-clientReadBufferSize_-1-clientWriteBufferSize_-1-serverReadBufferSize_-1-serverWriteBufferSize_0-:
50_Latency: 321.7920µs	90_Latency: 614.2500µs	99_Latency: 929.6250µs	Avg_Latency: 440.0230µs	Bytes/op: 27984.70085846357	Allocs/op: 191.01549636803875	
Histogram (unit: µs)
Count: 22715  Min:  61.8  Max: 101223.3  Avg: 440.02
------------------------------------------------------------
[       61.791000,        61.792000)      1    0.0%    0.0%  
[       61.792000,        61.798753)      0    0.0%    0.0%  
[       61.798753,        61.851102)      0    0.0%    0.0%  
[       61.851102,        62.256949)      0    0.0%    0.0%  
[       62.256949,        65.403307)      1    0.0%    0.0%  
[       65.403307,        89.795692)     67    0.3%    0.3%  
[       89.795692,       278.899567)   8888   39.1%   39.4%  ####
[      278.899567,      1744.942159)  13737   60.5%   99.9%  ######
[     1744.942159,     13110.552117)      1    0.0%   99.9%  
[    13110.552117,    101223.333000)     19    0.1%  100.0%  
[   101223.333000,    784324.564182)      1    0.0%  100.0%  
Number of requests:  22715	Request throughput:  7.4432512e+08 bit/s
Number of responses: 22715	Response throughput: 0 bit/s

go1.19.1/grpc1.51.0-dev
unary-networkMode_none-bufConn_false-keepalive_false-benchTime_1s-trace_false-latency_0s-kbps_0-MTU_0-maxConcurrentCalls_10-reqSize_4096B-respSize_0B-compressor_off-channelz_false-preloader_false-clientReadBufferSize_-1-clientWriteBufferSize_-1-serverReadBufferSize_-1-serverWriteBufferSize_65536-:
50_Latency: 215.9160µs	90_Latency: 503.0830µs	99_Latency: 789.1670µs	Avg_Latency: 296.4890µs	Bytes/op: 27760.243928507807	Allocs/op: 169.26183124517547	
Histogram (unit: µs)
Count: 33682  Min:  42.2  Max: 101288.1  Avg: 296.49
------------------------------------------------------------
[       42.250000,        42.251000)      1    0.0%    0.0%  
[       42.251000,        42.257753)      0    0.0%    0.0%  
[       42.257753,        42.310114)      0    0.0%    0.0%  
[       42.310114,        42.716079)      0    0.0%    0.0%  
[       42.716079,        45.863645)      5    0.0%    0.0%  
[       45.863645,        70.267660)    471    1.4%    1.4%  
[       70.267660,       259.479211)  19818   58.8%   60.3%  ######
[      259.479211,      1726.492400)  13377   39.7%  100.0%  ####
[     1726.492400,     13100.680054)      0    0.0%  100.0%  
[    13100.680054,    101288.125000)      9    0.0%  100.0%  
[   101288.125000,    785031.499244)      1    0.0%  100.0%  
Number of requests:  33682	Request throughput:  1.103691776e+09 bit/s
Number of responses: 33682	Response throughput: 0 bit/s
```

> On another note, I just realized that we are not sanitizing the buffer sizes received through the dial options. Passing negative values would cause a panic I think. So, I would advise against using a negative value for the flag to indicate the use of the default value.

If any buffer size is set to a negative value, at the moment it does not panic but disables buffering by:
- reseting it to 0 on the write side
- not using a `bufio.Reader` on the read side
See https://github.com/grpc/grpc-go/blob/689d061d46c76f52f6a965ef621f988f15655200/internal/transport/http_util.go#L365-L371.

I assumed that we want a way to run the benchmark with the default value, but all positive values are valid. What do you suggest instead? Do you think it is confusing that negative values have a different meaning in benchmarks and in `DialOption` or `ServerOption`?

Reply:
Yes, I feel that if the flag is not specified, we should use the default. If the user of the benchmark wants to explicitly disable buffering, they can set the value of the flag to `0`. What do you think about that approach?

Also, I realized that the documentation for these options can be clarified to say that negative values will disable buffering. I will file an issue to track that separately. Thank you.

Reply:
> Also, I realized that the documentation for these options can be clarified to say that negative values will disable buffering. I will file an issue to track that separately. Thank you.

Thanks. (this is being done here https://github.com/grpc/grpc-go/pull/580)

> Yes, I feel that if the flag is not specified, we should use the default. If the user of the benchmark wants to explicitly disable buffering, they can set the value of the flag to 0. What do you think about that approach?

I like it on paper, but in practice this requires encoding the information about using the default as a separate features (for example a `UseDefaultClientWriteBufferSize` boolean), and then check whether there were explicit flags to set it. This would make the `Feature` struct larger, which has its downside (it's currently being serialized on each test to see what's beeing testing). You could still argue that it's simpler/more explicit. Is that what you have in mind?

Reply:
I was thinking more like what we do for some of the other options like `maxConcurrentCalls` or `reqSizeBytes` or `respSizeBytes`, where we have a bunch of defaults:

```
	defaultMaxConcurrentCalls = []int{1, 8, 64, 512}
	defaultReqSizeBytes       = []int{1, 1024, 1024 * 1024}
	defaultRespSizeBytes      = []int{1, 1024, 1024 * 1024}
```

and we set the default value of the flag to these defaults.

Seriously though, this benchmark code is so convoluted.

Reply:
`maxConcurrentCalls`, `reqSizeBytes` and `defaultRespSizeBytes` are properties of the benchmark, rather than settings exposed by the libraries. For this reason, there is no obvious defaults for them and what the benchmark defines is somewhat arbitrary.

In contrast, for settings like read/write buffer sizes, in most cases we should benchmark using the default value provided by the library itself, defined here: https://github.com/grpc/grpc-go/blob/master/clientconn.go#L99-L100, unless we're trying to tune those settings. The read/write buffer size constants are not public, so we can't access them from the benchmark code. We need a way to say "don't set read/write buffer size explicitly" in the benchmark features, hence my suggestion of using a special value (-1) or a boolean feature for this.

The alternative you suggest would require either 
1. making `defaultWriteBufSize` and `defaultReadBufSize` public to use it in the benchmark, or
2. copying its value in the benchmark.

Either way since the settings will now be explicit, if we ever change the default in the options to e.g. dynamic buffer sizes this will need to be reflected in the benchmarks by changing the defaults. By contrast if we have a value that indicates "don't set anything, use the default behaviour" we're guaranteed that the benchmarks use the default library behavior. Maybe I'm overthinking this?

Reply:
Makes sense. Thank you.